src/vfs2.c

```diff
@@ -39,119 +40,241 @@
-|                                   EMPTY                                    |                           |         FLUSH          |  |
-|                                                                            | <------------------------ |                        |  |
-+----------------------------------------------------------------------------+   sqlite3_wal_checkpoint  +------------------------+  |
-  |                       ^                                                                                |                         |
-  | vfs2_commit_barrier   | sqlite3_wal_checkpoint                                                         |                         |
-  v                       |                                                                                |                         |
-+----------------------------------------------------------------------------+  vfs2_commit_barrier        |                         |
-|                                    BASE                                    | <---------------------------+                         |
-+----------------------------------------------------------------------------+                                                       |
-  |                       ^                          |                                                                               |
-  | sqlite3_step          | vfs2_unhide              +-------------------------------------------------------------------------------+
-  v                       |
-+----------------------+  |
-|        ACTIVE        |  |
-+----------------------+  |
-  |                       |
-  | COMMIT_PHASETWO       |
-  v                       |
-+----------------------+  |
-|        HIDDEN        |  |
-+----------------------+  |
-  |                       |
-  | vfs2_poll             |
-  v                       |
-+----------------------+  |
-|        POLLED        | -+
-+----------------------+
-
-*/
-
-static const struct sm_conf wtx_states[SM_STATES_MAX] = {
+/**
+ * Major transitions for this state machine:
+ *
+ *     +-----------CLOSED----------+
+ *     |                           |
+ * xOpen && no tx in WAL-cur   xOpen && tx in WAL-cur
+ *     |                           |
+ *     |                 (FB)      |
+ *     v      (BE)     <------     v  v-+ vfs2_add_uncommitted
+ *     EMPTY<------BASE------>FOLLOWING-+
+ *                 ^  |  (BF)
+ *                 |  |
+ *                 |  | sqlite3_step && xWrite(WAL)
+ *                 |  |
+ *                 |  v    v-+ xWrite(WAL)
+ *                 |--ACTIVE-+
+ *                 |  |
+ *      vfs2_abort |  | COMMIT_PHASETWO
+ *                 |  |
+ *                 |  v
+ *                 |--HIDDEN
+ *                 |  |
+ *                 |  | vfs2_poll
+ *                 |  |
+ *                 |  v
+ *                 +--POLLED
```

Comment:
Can't you move from POLLED back to BASE without calling `vfs2_abort`?

Reply:
Yep, forgot to mention vfs2_unhide!